### ISSUE TITLE
optimize_read_in_order does not work on Merge Table

### ISSUE THREAD

**Issue by cw9** [ClickHouse/ClickHouse#7617]
Thank you guys for introducing the performance improvement with `optimize_read_in_order`, it works very well on ReplicatedMergeTree tables. 

When I create Distributed table on top of the ReplicatedMergeTree table, the performance is also there. But in my test when I select through a Merge table that is built on ReplicatedMergeTree tables with the same schema, the performance improvement is gone, it's scanning the full table again.

Is this expected?

**Comment by den-crane** [comment]
Do you mean Engine=Merge? 
Do you reproducible example?

**Comment by cw9** [comment]
Yes, here is the example:


create table ORX( A Int64, B Int64, V Int64) Engine=MergeTree order by (A,B)

create table m as ORX ENGINE = Merge(default, 'ORX')

insert into ORX select 111, number, number from numbers(10000000)

select * from ORX where A = 111 order by A,B limit 3 settings max_threads=1;
```
3 rows in set. Elapsed: 0.038 sec. Processed 245.76 thousand rows, 5.90 MB (6.52 million rows/s., 156.57 MB/s.)
```

select * from m  where A = 111 order by A,B limit 3 settings max_threads=1;
```
3 rows in set. Elapsed: 0.469 sec. Processed 10.00 million rows, 240.00 MB (21.34 million rows/s., 512.17 MB/s.)
```

**Comment by blinkov** [comment]
cc @CurtizJ

**Comment by cw9** [comment]
Hi, is there any update on this issue?

**Comment by CurtizJ** [comment]
This behavior is expected, optimization for Merge tables wasn't planned. However, it's possible to implement optimization for source tables of Merge table, that have MergeTree-family engine, but it requires some refactoring. Now optimization is performed at query analyzation stage, before Merge table asks it's source tables to read data and we don't know about their structure. In distributed tables, query is sent at each node and runs there independently, so optimization works in that case.
So, this optimization request isn't declined, but isn't first priority.

**Comment by CurtizJ** [comment]
Implemented in master #8130.

**Comment by cw9** [comment]
Looks like there is a regression on this? 
```
create table mt( A Int64, B Int64, V Int64) Engine=MergeTree order by (A,-B)
insert into mt select 111, number, number from numbers(10000000)

select * from mt where A = 111 order by A,-B limit 3 settings max_threads=1,optimize_read_in_order=1;
┌───A─┬───────B─┬───────V─┐
│ 111 │ 9999999 │ 9999999 │
│ 111 │ 9999998 │ 9999998 │
│ 111 │ 9999997 │ 9999997 │
└─────┴─────────┴─────────┘

3 rows in set. Elapsed: 0.289 sec. Processed 10.00 million rows, 240.00 MB (34.55 million rows/s., 829.17 MB/s.)
```
I expect the select query to read less than 10M rows
